### PR TITLE
Creating `DrawPanel` UI

### PR DESCRIPTION
**Pull request by ganimtron-10**
This UI would help us to create various shapes and transform them.

Currently, it supports
- Creation of
  - Line
  - Circle
  - Quad
- Selection (Translation)
- Deletion
- Shapes are perfectly clamped

Demo

python_Dg9yGkR9DT

### CODE REVIEW

File: fury/ui/elements.py
Comment (skoudoro) [inline fury/ui/elements.py:3074]: please, initialize `self.shape = None`. talking about that, `self.shape` should be renamed (e.g pencil, pen, tool, paint, brush, etc.... I let you pick)

File: fury/ui/elements.py
Comment (skoudoro) [inline fury/ui/elements.py:3124]: please add a docstring

File: fury/ui/elements.py
Comment (skoudoro) [inline fury/ui/elements.py:3141]: I think you should work with the whole NumPy array and avoid this loop.

- add a docstring
- you can delete `rotate_point`
- use `utils.vertices_from_actor` to get all points as a NumPy array
- do your rotation
- use `numpy_to_vtk_points` + SetPoint or `utils.set_polydata_vertices`

File: fury/ui/elements.py
Comment (skoudoro) [inline fury/ui/elements.py:3179]: `,optional` to add

**Review by skoudoro — CHANGES_REQUESTED** [review]
Hi @ganimtron-10,

Thank you for this feature. This is really nice! see below for some comments.

it would be also good :
- if you could add the erase functionality
- To add an option to tell if `DrawPanel` is movable or not
- When we start, we see mode selection, but nothing happens when we click the drawer panel, it is confusing. it will be good to do something. 
- it would be good to use the on/off capabilities of button2D to know which tool we are using

**Comment by m-agour** [comment]
Hello @ganimtron-10 ,
Great job you did there! It's really nice to draw primitives like that, maybe for describing a scene or draw an arrow to point at something and a lot of other useful applications.

- I experimented with the panel, and I have found some minor glitches:

1. when I click in place in drawing mode, it draws a time point which does not get rendered until I draw a big shape or force render the scene somehow (I'm not sure if it should be allowed to draw shapes in place unless I'm drawing a point).

https://user-images.githubusercontent.com/63170874/172949975-6d4f3cc5-08f8-4417-9816-db46da5453f0.mp4

2. The above issue also triggered the below runtime warning:

```
 ..\fury\fury\ui\elements.py:3149: RuntimeWarning: divide by zero encountered in long_scalars
  self.rotate_line(deg=np.arctan(size[1]/size[0]))
 ```

- Another minor note is there should be a separate selection mode and selecting shapes while drawing should be disabled or handled in another way, because in the current state I can't draw shapes above shapes.

File: fury/ui/elements.py
Comment (m-agour) [reply fury/ui/elements.py:3149]: +1. I suggest you use np.arctan2 which takes care of choosing the correct angle.

File: fury/ui/elements.py
Comment (ganimtron-10) [reply fury/ui/elements.py:3060]: But this UI only creates and transforms the shape depending on `shape_type`, It doesn't have any `Panel2D`.

File: fury/ui/elements.py
Comment (ganimtron-10) [reply fury/ui/elements.py:3220]: Here in the setup, the `self.current_mode` isn't accessible as we have to change the text of `TextBlock2D` whenever the mode is changed, which is done after the creation of `TextBlock2D`.
I think it would be great to keep "Mode = " (blank) at the start because there isn't any significance of selection mode.

File: fury/ui/elements.py
Comment (xtanion) [inline fury/ui/elements.py:3149]: I encountered a few RuntimeErrors in this line:
```
RuntimeWarning: invalid value encountered in long_scalars
  self.rotate_line(deg=np.arctan(size[1]/size[0]))
```
Also, I faced the same RuntimeError due to division by 0.

**Review by xtanion — COMMENTED** [review]
Hi @ganimtron-10, This feature is really nice and everything is working fine. There are a few runtime warnings I encountered, Look at my comment below.
Thanks.

File: fury/ui/elements.py
Comment (ganimtron-10) [reply fury/ui/elements.py]: Solved 👍

File: fury/ui/elements.py
Comment (ganimtron-10) [reply fury/ui/elements.py]: Solved 👍

File: fury/ui/elements.py
Comment (ganimtron-10) [reply fury/ui/elements.py:3291]: If the `Shape2D` gets an appropriate name then we could rename every related function accordingly.

**Comment by ganimtron-10** [comment]
Thank You @m-agour and @xtanion for your wonderful review!!

File: fury/ui/elements.py
Comment (ganimtron-10) [reply fury/ui/elements.py]: This property will hold a reference to an `Actor2D`, which is then added to the `DrawPanel`.
As I consider elements on `DrawPanel` as shapes, I have named it the same.

File: fury/ui/elements.py
Comment (ganimtron-10) [reply fury/ui/elements.py]: Done 👍

File: fury/ui/elements.py
Comment (ganimtron-10) [reply fury/ui/elements.py]: Added "Select appropriate drawing mode using below icon" message on Startup.

**Comment by ganimtron-10** [comment]
Hello @skoudoro ,
I have made all the proposed changes except the renaming part, it's still confusing as the name suggested are not clearly depicting the work they do in my opinion. Maybe we would have to discuss this.

> * it would be good to use the on/off capabilities of button2D to know which tool we are using

Regarding this, there aren't any on/off capabilities in `Button2D`, it is present in `Option` and `Checkbox` UI.
Also to know in which mode we are currently in, there's a Label at the top or to make it more intuitive we can work on the icons and create two separate icon sets such as static and pressed and show them accordingly.

Thanks!

File: fury/ui/elements.py
Comment (ganimtron-10) [reply fury/ui/elements.py:3059]: Renamed the class to `DrawShape`.

**Comment by ganimtron-10** [comment]
To show which mode is currently selected, I have made some quick icons by modifying the existing ones. (These are just for testing purposes)

Demo
DrawPanel-Update Icon

To try the above changes follow the below steps:
1. Download [this](https://github.com/fury-gl/fury/files/8898824/new_icon.zip) file.
2. Extract the file and put the icons in the `\.fury\icons\icomoon` folder.
3. Then execute the tutorial.

File: fury/ui/elements.py
Comment (skoudoro) [inline fury/ui/elements.py:3191]: `, optional` to add

File: fury/ui/elements.py
Comment (skoudoro) [inline fury/ui/elements.py:3195]: The first time, I did not succeed to drag the panel despite this option to True.

But after looking at the code, I see that I need to be in "selection" mode to make it work. This is cool but can be confusing. people will not look at the code to guess how to use it.

File: fury/ui/elements.py
Comment (skoudoro) [inline fury/ui/elements.py:3223]: this size should be in the init function, we might want to make it smaller. That is also why mode_data should be a property

File: fury/ui/elements.py
Comment (skoudoro) [inline fury/ui/elements.py:3214]: this `mode_data`, should be a private variable (`self._mode_data`). Then you create a property `self.mode_data` (read-only first in this PR, but in the close future, the user can add a mode easily by adding it in the dictionary)

File: fury/ui/elements.py
Comment (skoudoro) [inline fury/ui/elements.py:3293]: Can you explain why you need to force render here ?

File: fury/ui/elements.py
Comment (skoudoro) [inline fury/ui/elements.py:3087]: Why do we use a rectangle2D for the line and not a line actor?

File: fury/ui/elements.py
Comment (ganimtron-10) [reply fury/ui/elements.py:3293]: Whenever the current mode was changed, the message in the`TextBlock2D` was also updated to show the current mode, at that time until the `force_render()` method wasn't called the new mode wasn't displayed (because internally `TextBlock2D` doesn't render whenever the message changes) So I had to do it manually.

But now as we update the button icons simultaneously, Now we don't need it.

File: fury/data/fetcher.py
Comment (m-agour) [inline fury/data/fetcher.py:29]: Hello @ganimtron-10, 
These icons shouldn't be in your fork of the repo but rather PR them to the main repo.

Comment (ganimtron-10) [reply fury/data/fetcher.py:29]: How the icons should be fetched was not decided yet!
Either we merge the old icons with new ones and put them in fury-data or
we create separate fetchers for both of them.
So I decided to use my fork repo for now and update it accordingly.

Comment (m-agour) [reply fury/data/fetcher.py:29]: Yes, sorry I thought you finished with this PR. Great job btw.

**Comment by ganimtron-10** [comment]
Hello @Garyfallidis @nasimanousheh @skoudoro ,
I have updated the `DrawPanel` and `DrawShape` UI, due to which the shapes are now restricted to remain inside the canvas.
PTAL.

The tests are failing because I have separated the code for the [fetcher function](https://github.com/fury-gl/fury/pull/609) which fetches new icons required for the `DrawPanel UI`.

Thanks!

File: fury/ui/elements.py
Comment (skoudoro) [inline fury/ui/elements.py:21]: pep8? `set_input` not used

File: fury/ui/elements.py
Comment (skoudoro) [reply fury/ui/elements.py:3272]: can you note this somewhere as a future new feature

File: fury/ui/elements.py
Comment (skoudoro) [reply fury/ui/elements.py:3281]: can you note this somewhere as a future new feature

File: fury/ui/elements.py
Comment (skoudoro) [inline fury/ui/elements.py:3401]: `last_mode`?  I do not see it

File: fury/ui/elements.py
Comment (skoudoro) [inline fury/ui/elements.py:3409]: ==1 ? what does it mean?

File: fury/ui/elements.py
Comment (skoudoro) [inline fury/ui/elements.py:3380]: Not sure of this variable goal... Can you explain?

**Review by skoudoro — COMMENTED** [review]
Hi @ganimtron-10,

This first version looks really good. 

we can not test this tutorial because the fetcher is missing. Since the fetcher PR has been merged, Can you merge master or rebase this PR?

After that, and few other fixes, it should be ready to be merged. Thanks for this clean work

File: fury/ui/elements.py
Comment (ganimtron-10) [reply fury/ui/elements.py:3380]: When the `left_mouse_button` is pressed we initiate the drawing of the shape (ie. add the starting point of the shape) then as we want to resize the shape according to the mouse movement, I need to check whether the shape creation is in the process to apply the resizing according to the current mouse position.

File: fury/ui/elements.py
Comment (ganimtron-10) [reply fury/ui/elements.py:3409]: These buttons basically have two icons 
0th index/id - static icon
1st index/id - pressed icon
So I am updating the icon of the `current_mode` button 
and checking if any other button has its icon set to the pressed mode (which is 1st index/id) then updating it too.

Comment (skoudoro) [reply fury/ui/elements.py:3408]: make sense. thanks!

**Review by skoudoro — APPROVED** [review]
ok, LGTM! thanks @ganimtron-10!

This is ready to go. I will wait for another review until Wednesday. if no other review, I will go ahead and merge it.

I found one small bug but I think, this fix or any new features should be done in a new PR.

Good job!

**Review by nasimanousheh — APPROVED** [review]

**Comment by nasimanousheh** [comment]
Hi @ganimtron-10. Great job! you fixed the boundary issue and by moving the shape, it doesn't go out of the canvas boundary. 
I found two bugs in your recent changes:
- I can not move the shape with one click. I need to double click on the shape to move it.
- When I double click on the shape, without moving the mouse, the shape jumps to other position. As you see here:

fail

File: fury/ui/elements.py
Comment (ganimtron-10) [reply fury/ui/elements.py:3371]: Maybe I should rename this as `is_drawing` because we are basically checking whether the shape is being drawn.
The job of this parameter is to indicate whether the shape is to be drawn (`in_process` is false viz. shape drawing is not in process) or the shape is already drawn we just have to resize it (`in_progress` is true viz. the shape drawing is in process).

Also, you are getting the _IndexError_ because the shape isn't created yet. As we are saying the shape creation is in process (`in_process` = True) there should exist an actor in the `actor_list` which is not in this case.

**Comment by ganimtron-10** [comment]
Hello @nasimanousheh ,


> * I can not move the shape with one click. I need to double click on the shape to move it.

When I use my external mouse single click with the dragging works pretty fine, but when I use my trackpad mouse I have to double click and then drag.



> * When I double click on the shape, without moving the mouse, the shape jumps to other position. As you see here:
> 
> fail 

There was an issue with the positioning of the element in the panel which is now fixed using relative positioning.

PTAL.

Thanks!

File: fury/ui/elements.py
Comment (nasimanousheh) [inline fury/ui/elements.py:3284]: You can include the fourth variable for color 'rgba'.

File: fury/ui/elements.py
Comment (ganimtron-10) [reply fury/ui/elements.py:3149]: The `bounding_box` I am calculating here is the white-bordered box which is shown below
image

File: fury/ui/elements.py
Comment (ganimtron-10) [reply fury/ui/elements.py:3284]: The alpha value is explicitly set using the `opacity` parameter for all the UI

File: fury/ui/elements.py
Comment (ganimtron-10) [reply fury/ui/elements.py:3162]: As we are just considering numeric values it won't affect the calculation as the same value will get updated if we use `<=`.

**Comment by Garyfallidis** [comment]
Great job @ganimtron-10 ! Merging. Bring forth the titans aka next PRs!

**Comment by ganimtron-10** [comment]
python_cxQFjGZBM9

**Pull request merged**
